app/controllers/profiles_controller.rb

```diff
@@ -14,51 +14,15 @@ class ProfilesController < ApplicationController
 
   def index
     if params[:search]
-      @profiles = profiles_for_search
-      # sum of search results concerning certain attributes
-      @aggs = @profiles.response.aggregations
-      @aggs_languages = @aggs[:lang][:buckets]
-      @aggs_cities = @aggs[:city][:buckets]
-      @aggs_states = @aggs[:state][:buckets]
-      @aggs_countries = @aggs[:country][:buckets]
-    elsif params[:tag_filter]&.present?
-      @tags = params[:tag_filter].split(/\s*,\s*/)
-      @profiles = profiles_with_tags
-      # redirect_to profiles_path(:anchor => "speakers")
-    elsif params[:category_id]
-      @profiles = profiles_for_category
+      search_with_search_params
     else
-      @profiles = profiles_for_index
-    end
-
-    @profiles_count = @profiles.total_count
-
-    # for the tags filter module that is available all the time at the profile index view
-    # is needed for the colors of the tags
-    @category =
-      if params[:category_id]
-        Category.find(params[:category_id])
-      elsif params[:tag_filter]
-        if params[:tag_filter].empty?
-          redirect_to profiles_url(anchor: "top"), notice: I18n.t('flash.profiles.no_tags_selected')
-          return
-        end
-        last_tag = params[:tag_filter].split(/\s*,\s*/).last
-        last_tag_id = ActsAsTaggableOn::Tag.where(name: last_tag).last.id
-        Category.select{|cat| cat.tag_ids.include?(last_tag_id)}.last
-      else
-        Category.first
+      if params[:tag_filter]&.empty? 
+        redirect_to profiles_url(anchor: "top"), notice: I18n.t('flash.profiles.no_tags_selected')
+        return 
       end
-    @categories = Category.sorted_categories
-    Category.all.includes(:translations).each do |category|
-      tags = ActsAsTaggableOn::Tag
-        .belongs_to_category(category.id)
-        .with_published_profile
-        .with_regional_profile(search_region)
-        .translated_in_current_language_and_not_translated(I18n.locale)
-      tags = tags.belongs_to_more_than_one_profile unless current_region
-      instance_variable_set("@tags_#{category.short_name}", tags).most_used(100)
-    end
+      search_with_categories_and_tags
```

Comment:
Since there are still 4 different cases (no params, `search` param, `tag_filter` param and `category_id` param), I think it'll be more readable, if all 4 cases were visible in this method, instead of just having `search` here, and the rest in `search_with_categories_and_tags`. You can still extract some common functionality and put it in some method (e.g. `build_categories_and_tags`, and call it in the relevant cases.

Reply:
thanks for the feedback. how is it looks now?